csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -137,43 +140,37 @@ public ValueTask SetAdapterDirectProxyAsync(string adapterId, IObjectPrx? proxy,
                     return null;
                 }
 
-                IObjectPrx prx = _wellKnownProxy.Clone(IObjectPrx.Factory, identity: identity);
-
-                var adapterIds = new List<string>();
-                foreach (KeyValuePair<string, HashSet<string>> entry in _replicaGroups)
+                foreach ((string key, HashSet<string> ids) in _replicaGroups)
```

Comment:
previously we randomly return a replica group, but can the object be hosted by several replica-groups? I think an object can be part of a replica-group or be hosted on an adapter that is not part of a replica group, but it cannot be hosted by several replica-groups or multiple adapters that are not part of the same replica-group

Reply:
We should at the very least return a proxy that points to the replica group and not a proxy that points on the first member of the replica group.

IceGrid disallow registration of a well-known object with different replica groups/adapters. 

For IceDiscovery, there's no explicit registration and instead it allowed resolving a replicated object from replica groups or adapters with different IDs. While probably uncommon to host a well-known object in servers with different replica groups, it could be more common for servers with different adapters and I don't really see a good reason to disallow this.

Reply:
>We should at the very least return a proxy that points to the replica group and not a proxy that points on the first member of the replica group.

```
IObjectPrx proxy = _adapters[ids.First()];
proxy = proxy.Clone(IObjectPrx.Factory, adapterId: key, identity: identity);
```
Isn't that what this code does? we are cloning the proxy using the replica-group as the adapterId 


> For IceDiscovery, there's no explicit registration and instead, it allowed resolving a replicated object from replica groups or adapters with different IDs. While probably uncommon to host a well-known object in servers with different replica groups, it could be more common for servers with different adapters and I don't really see a good reason to disallow this.

I fine with keeping the previous behavior, but we should refactor this code to collect the proxies and the return a random one, as each adapter, replica-group can use a different protocol

Reply:
Yes, you're right I misread the proxy clone call and thought it would return a proxy on the first OA.

Reply:
So in conclusion, the randomness is gone because it's very uncommon for the same object to be hosted in multiple replica groups or non-replicated object adapters.